agent/lib_mongodb.c

```diff
@@ -173,7 +175,205 @@ NR_PHP_WRAPPER(nr_mongodb_operation) {
 }
 NR_PHP_WRAPPER_END
 
-void nr_mongodb_enable(TSRMLS_D) {
+#else
+
+NR_PHP_WRAPPER(nr_mongodb_operation_before) {
+    (void)wraprec;
+    nr_segment_t* segment;
+    segment = nr_segment_start(NRPRG(txn), NULL, NULL);
+    segment->wraprec = auto_segment->wraprec;
+}
+NR_PHP_WRAPPER_END
+
+NR_PHP_WRAPPER(nr_mongodb_operation_after) {
+  const char* this_klass = "MongoDB\\Operation\\Executable";
+  zval* collection = NULL;
+  zval* database = NULL;
+  zval* server = NULL;
+  zval* this_var = NULL;
+  nr_datastore_instance_t instance = {
+      .host = NULL,
+      .port_path_or_id = NULL,
+      .database_name = NULL,
+  };
+  nr_segment_datastore_params_t params = {
+    .datastore = {
+      .type = NR_DATASTORE_MONGODB,
+    },
+    .operation = nr_strdup (wraprec->extra),
+    .instance = &instance,
+    .callbacks = {
+      .backtrace = nr_php_backtrace_callback,
+    },
+  };
+
+  /*
+   * We check for the interface all Collection operations extend, rather than
+   * their specific class. Not all operations have the properties we need but
+   * the ones we hook do (as of mongo-php-library v.1.1).
+   */
+  this_var = nr_php_scope_get(NR_EXECUTE_ORIG_ARGS);
+  if (!nr_php_object_instanceof_class(this_var, this_klass)) {
+    nrl_verbosedebug(NRL_FRAMEWORK, "%s: operation is not %s", __func__,
+                     this_klass);
+    goto leave;
+  }
+
+  collection
+      = nr_php_get_zval_object_property(this_var, "collectionName");
+  if (nr_php_is_zval_valid_string(collection)) {
+    params.collection = Z_STRVAL_P(collection);
+  }
+
+  database
+      = nr_php_get_zval_object_property(this_var, "databaseName");
+  if (nr_php_is_zval_valid_string(database)) {
+    instance.database_name = Z_STRVAL_P(database);
+  }
+
+  server = nr_php_arg_get(1, NR_EXECUTE_ORIG_ARGS);
+  nr_mongodb_get_host_and_port_path_or_id(server, &instance.host,
+                                          &instance.port_path_or_id);
+
+  nr_segment_datastore_end(&auto_segment, &params);
+
+leave:
```

Comment:
If we `goto leave`, we end up with a dangling segment. This shouldn't be too bad, as the segment storage is within a transaction, so when the transaction ends the segment is cleaned up. Do we want to end the segment with limited information if we exit early?

I am dealing with this exact question with our drupal instrumentation right now

Reply:
addressed here 57364ba6937aef21bacb8c2e885eb670b32b796c